.changeset/ten-cougars-pretend.md

```diff
@@ -0,0 +1,87 @@
+---
+'@astrojs/netlify': major
+---
+
+# Netlify Adapter v4 simplifies static + SSR deployments
+
+This update is a complete overhaul of the Netlify adapter.
+It simplifies the user-facing config, and resolves a number of bugs along the way.
+
+Here's what changes:
+
+## Middleware without configuration
+
+In v3, [Middleware](https://docs.astro.build/en/guides/middleware/) was deployed to the SSR Function by default.
+Enabling the `edgeMiddleware` option would change this behaviour to deploy it using Netlify Edge Functions instead, improving
+page speed and enabling the middleware to run on prerendered sites.
+
+In v4, this is the default setting.
+If your Astro site makes use of Middleware,
+it's deployed to an Edge Function where it runs super close to your users and is also applied to static sites.
+You can use this to implement customization or authorization logic!
+
+**Action Required:**
+If your site is using Middleware and you had `edgeMiddleware` enabled, nothing changes for you. You are free to remove the config field, as it is no longer necessary.
+If your site is using Middleware and you didn't have `edgeMiddleware` enabled, your Middleware will now be deployed to an Edge Function and start running on prerendered pages. Go through your prerendered pages to check if this requires any changes to the logic in your middleware.
+
+## Netlify Context is automatically available via Locals
+
+In v3, you could use `netlify-edge-middleware.ts` to inject data from the Netlify context into your Astro locals.
+In v4, this file is no longer needed because the Netlify context is automatically made available via `Astro.locals.netlify.context`.
+You can use this context to access information about the user (like geolocation or IP address), your Netlify site (like deploy ID) or the request (like its request ID or the CDN region it's served from).
+
+**Action Required:**
+Remove the `netlify-edge-middleware.ts` or `netlify-edge-middleware.js` file.
+In your codebase, change all usage of locals injected through that file to use `Astro.locals.netlify.context` instead.
+
+### Image CDN
+
+v4 of this adapter integrates your Astro site with Netlify [Image CDN](https://docs.netlify.com/image-cdn/overview/).
+This allows transforming images on-the-fly without impacting build times.
+It's implemented using an [Astro Image Service](https://docs.astro.build/en/reference/image-service-reference/), and enabled by default.
+
+## On-Demand Builders are no longer supported
```

Comment:
On-demand builders have a special place in the astro community.

The vast majority of our users are concerned with static sites. When their sites become large, our users start looking around for options to reduce work done during build-time but are intimidated by the world of complexity that SSR introduces.

With that context, when I introduce builders to them and they ask, "Is it really as simple as enabling a config?", I have been able to say yes without elaboration.

Reply:
It does not sound like an issue sitting here at all, but "how much do I need to learn" is a thought users always have in my experience, and tiny things like this are the ones to make them decide one way or another.

Reply:
I am not sure.

Personally, I'd want on-demand builders to stay. I would have removed the lambda compatible API in my PR too if the new API had a "persistent" mode.

We could add things by default but that would just be hiding complexity. If the stance from Netlify really is to remove builders at some point, the users will just have to read up more on server stuff and that's fine too.

Just meant to point out that this is bit of a downgrade in one aspect.

Reply:
> If the stance from Netlify really is to remove builders at some point

I don't see that happening anytime soon - we have happy customers relying on builders, and we'll keep supporting that. But we consciously haven't added support for it to V2 Functions, because we see the cache headers as the new thing. But while they're great tools for framework implementors, I agree they're pretty complex things for most framework users.

What I think we should do us to add a `cache` option to the Adapter config. It would serve as a replacement for the `builders` option that currently exists.
If the user sets `cache: true`, the generated code will apply some sane `CDN-Cache-Control` header that behaves similar to builders. If the user wants to customize it per-page, like with `setBuildersTtl`, they can override it by setting Cache-Control headers from within the page itself.

Reply:
That sounds perfect.

One thing I am curious about is how long far-future cache stays. As a demo for ODM, I often share this: https://github.com/lilnasy/astro-netlify-isr. The rendered date of the live demo is for July. It is easy to explain it in terms of "static but built later". But I'm guessing we would need to document that far-future cache may be evicted sooner?

Reply:
Correct, that cache is evicted once the TTL elapses, or the CDN drops the cache for some other reason. So it's not as obvious as with ODBs when it's regenerated. But developers won't really need to think about this, so I don't think that needs to be explained in detail in docs.

I'll get to work on the `cache` option 👍